File: functions/src/services/knowledgeService.ts
import { db, EMBEDDING_DIMENSION, EMBEDDING_MODEL } from '../config.js';
import { getGenAiClient } from './genaiClient.js';

export interface ChunkResult {
    content: string;
    metadata: any;
    similarity: number;
}

function normalizeEmbedding(embedding: number[]): number[] {
    const norm = Math.sqrt(embedding.reduce((sum, val) => sum + val * val, 0));
    if (norm === 0) return embedding;
    return embedding.map(val => val / norm);
}

export async function generateEmbedding(text: string, taskType: 'RETRIEVAL_QUERY' | 'RETRIEVAL_DOCUMENT'): Promise<number[]> {
    const ai = getGenAiClient();
    const embeddingDim = EMBEDDING_DIMENSION > 0 ? EMBEDDING_DIMENSION : undefined;
    const response = await ai.models.embedContent({
        model: EMBEDDING_MODEL,
        contents: text,
        config: {
            taskType,
            ...(embeddingDim ? { outputDimensionality: embeddingDim } : {}),
        },
    });

    return normalizeEmbedding(response.embeddings![0].values!);
}

export async function searchKnowledge(query: string, limit: number = 5): Promise<ChunkResult[]> {
    try {
        // 1. Generate embedding for the query
        const vector = await generateEmbedding(query, 'RETRIEVAL_QUERY');

        // 2. Search Firestore Vector Store
        // Note: Requires Firestore Vector Search index to be created
        const collection = db.collection('rag_chunks');
        const vectorQuery = collection.findNearest('embedding', vector, {
            limit,
            distanceMeasure: 'COSINE'
        });

        const snapshot = await vectorQuery.get();
        
        const results: ChunkResult[] = [];
        snapshot.forEach(doc => {
            const data = doc.data();
            // In some SDK versions, distance is automatically populated in doc.data() 
            // when using findNearest.
            const distance = data._distance ?? data.distance;
            results.push({
                content: data.content,
                metadata: data.metadata,
                similarity: typeof distance === 'number' ? 1 - distance : 0
            });
        });

        return results;

    } catch (error) {
        console.error('Vector search failed:', error);
        return [];
    }
}

File: functions-python/adk/knowledge_service.py
"""Knowledge base search (RAG) via Firestore vector search."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, List

import firebase_admin
from firebase_admin import firestore
from google.cloud.firestore_v1.base_vector_query import DistanceMeasure

from . import config
from .cache import TtlCache
from .genai_client import generate_embedding


def _ensure_firebase() -> None:
    if not firebase_admin._apps:
        firebase_admin.initialize_app()


@dataclass
class ChunkResult:
    content: str
    metadata: dict[str, Any]
    similarity: float


_cache = TtlCache[List[ChunkResult]](
    max_size=config.RAG_CACHE_MAX or 200,
    ttl_seconds=config.RAG_CACHE_TTL_SECONDS or 600,
)


def search_knowledge(query: str, limit: int = 5) -> List[ChunkResult]:
    normalized_query = query.strip().lower()
    if not normalized_query:
        return []

    cache_key = f"{normalized_query}:{limit}"
    cached = _cache.get(cache_key)
    if cached is not None:
        return cached

    _ensure_firebase()
    db = firestore.client()

    try:
        vector = generate_embedding(query, 'RETRIEVAL_QUERY')
        collection = db.collection('rag_chunks')
        vector_query = collection.find_nearest(
            'embedding',
            vector,
            limit=limit,
            distance_measure=DistanceMeasure.COSINE,
            distance_result_field='_distance',
        )
        snapshot = vector_query.get()

        results: List[ChunkResult] = []
        for doc in snapshot:
            data = doc.to_dict() or {}
            distance = data.get('_distance') or data.get('distance')
            similarity = 1 - float(distance) if distance is not None else 0.0
            results.append(
                ChunkResult(
                    content=data.get('content', ''),
                    metadata=data.get('metadata') or {},
                    similarity=similarity,
                )
            )

        _cache.set(cache_key, results)
        return results
    except Exception as exc:  # pragma: no cover - network/runtime dependent
        print(f'[RAG] search failed: {exc}')
        return []
